iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/util/DeviceRegistryTokenAuthProvider.java

```diff
@@ -36,17 +41,55 @@
 
     private final NamespacedKubernetesClient client;
     private final AuthApi authApi;
+    private final EmbeddedCacheManager cacheManager = new DefaultCacheManager();
+    private Cache<String, TokenReview> tokens;
+    private Cache<String, Boolean> authorizations;
 
-    public DeviceRegistryTokenAuthProvider() {
+    public DeviceRegistryTokenAuthProvider(Duration tokenExpiration) {
+        log.info("Using token cache expiration of {}", tokenExpiration);
         this.client = new DefaultKubernetesClient();
         this.authApi = new KubeAuthApi(this.client, this.client.getConfiguration().getOauthToken());
+        ConfigurationBuilder config = new ConfigurationBuilder();
+        config.expiration().lifespan(tokenExpiration.toSeconds(), TimeUnit.SECONDS);
+        cacheManager.defineConfiguration("tokens", config.build());
+        cacheManager.defineConfiguration("subjects", config.build());
+        tokens = cacheManager.getCache("tokens");
+        authorizations = cacheManager.getCache("subjects");
+
     }
 
     @Override
     public void authenticate(JsonObject authInfo, Handler<AsyncResult<User>> resultHandler) {
 
         final String token = authInfo.getString(TOKEN);
+        tokens.computeIfAbsentAsync(token, review -> authApi.performTokenReview(token))
+                .thenAccept(tokenReview -> {
+                    if (tokenReview != null && tokenReview.isAuthenticated()) {
+                        String role = getRole(authInfo);
+                        if (role != null) {
+                            RbacSecurityContext securityContext = new RbacSecurityContext(tokenReview, authApi, null);
+                            authorizations
+                                    .computeIfAbsentAsync(role, authorized -> securityContext.isUserInRole(role))
```

Comment:
This can again throw an Exception, in which case the handler wouldn't get called. Also, to my understanding, the method can block as it does a remote call. I think that would block the case. You could avoid that with `thenAcceptAsync`.

Reply:
As discussed offline, `computeIfAbsentAsync()` will execute blocking call async. And the exception will be handled by `exceptionally()` now.